controllers/app-node/chain/llmchain_controller.go

```diff
@@ -0,0 +1,139 @@
+		// There's no need to requeue if the resource no longer exists.
+		// Otherwise, we'll be requeued implicitly because we return an error.
+		log.V(1).Info("Failed to get LLMChain")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	log = log.WithValues("Generation", instance.GetGeneration(), "ObservedGeneration", instance.Status.ObservedGeneration, "creator", instance.Spec.Creator)
+	log.V(1).Info("Get LLMChain instance")
+
+	// Add a finalizer.Then, we can define some operations which should
+	// occur before the LLMChain to be deleted.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers
+	if newAdded := controllerutil.AddFinalizer(instance, arcadiav1alpha1.Finalizer); newAdded {
+		log.Info("Try to add Finalizer for LLMChain")
+		if err := r.Update(ctx, instance); err != nil {
+			log.Error(err, "Failed to update LLMChain to add finalizer, will try again later")
+			return ctrl.Result{}, err
+		}
+		log.Info("Adding Finalizer for LLMChain done")
+		return ctrl.Result{}, nil
+	}
+
+	// Check if the LLMChain instance is marked to be deleted, which is
+	// indicated by the deletion timestamp being set.
+	if instance.GetDeletionTimestamp() != nil && controllerutil.ContainsFinalizer(instance, arcadiav1alpha1.Finalizer) {
+		log.Info("Performing Finalizer Operations for LLMChain before delete CR")
+		// TODO perform the finalizer operations here, for example: remove vectorstore data?
+		log.Info("Removing Finalizer for LLMChain after successfully performing the operations")
+		controllerutil.RemoveFinalizer(instance, arcadiav1alpha1.Finalizer)
+		if err := r.Update(ctx, instance); err != nil {
+			log.Error(err, "Failed to remove the finalizer for LLMChain")
+			return ctrl.Result{}, err
+		}
+		log.Info("Remove LLMChain done")
+		return ctrl.Result{}, nil
+	}
+
+	instance, result, err := r.reconcile(ctx, log, instance)
+
+	// Update status after reconciliation.
+	if updateStatusErr := r.patchStatus(ctx, instance); updateStatusErr != nil {
+		log.Error(updateStatusErr, "unable to update status after reconciliation")
+		return ctrl.Result{Requeue: true}, updateStatusErr
+	}
+
+	return result, err
+}
+
+func (r *LLMChainReconciler) reconcile(ctx context.Context, log logr.Logger, instance *api.LLMChain) (*api.LLMChain, ctrl.Result, error) {
+	// Observe generation change
+	if instance.Status.ObservedGeneration != instance.Generation {
+		instance.Status.ObservedGeneration = instance.Generation
+		r.setCondition(instance, instance.Status.WaitingCompleteCondition()...)
+		if updateStatusErr := r.patchStatus(ctx, instance); updateStatusErr != nil {
+			log.Error(updateStatusErr, "unable to update status after generation update")
+			return instance, ctrl.Result{Requeue: true}, updateStatusErr
+		}
+	}
+
+	if instance.Status.IsReady() {
+		return instance, ctrl.Result{}, nil
+	}
+	// Note: should change here
+	if instance.Spec.Input.LLM.Name != "" && instance.Spec.Input.Prompt.Name != "" {
```

Comment:
we should do more checks later.For example:

- LLM status
- Prompt status

Reply:
Add TODO for this if we're not sure how to fix it for now.